mmgen/models/architectures/sngan_proj/generator_discriminator.py

```diff
@@ -0,0 +1,436 @@
+        # receive a noise generator and sample noise.
+        elif callable(noise):
+            noise_generator = noise
+            assert num_batches > 0
+            noise_batch = noise_generator((num_batches, self.noise_size))
+        # otherwise, we will adopt default noise sampler.
+        else:
+            assert num_batches > 0
+            noise_batch = torch.randn((num_batches, self.noise_size))
+
+        if isinstance(label, torch.Tensor):
+            assert label.ndim == 1, ('The label shoube be in shape of (n, )'
+                                     f'but got {label.shape}.')
+            label_batch = label
+        elif callable(label):
+            label_generator = label
+            assert num_batches > 0
+            label_batch = label_generator(num_batches)
+        elif self.num_classes == 0:
+            label_batch = None
+        else:
+            assert num_batches > 0
+            label_batch = torch.randint(0, self.num_classes, (num_batches, ))
+
+        # dirty code for putting data on the right device
+        noise_batch = noise_batch.to(get_module_device(self))
+        label_batch = label_batch.to(get_module_device(self))
+
+        x = self.noise2feat(noise_batch)
+        x = x.reshape(x.size(0), -1, self.input_scale, self.input_scale)
+
+        for conv_block in self.conv_blocks:
+            x = conv_block(x, label_batch)
+
+        out_feat = self.to_rgb(x)
+        out_img = self.final_act(out_feat)
+
+        if return_noise:
+            return dict(
+                fake_img=out_img, noise_batch=noise_batch, label=label_batch)
+        return out_img
+
+    def init_weights(self, pretrained=None, strict=True):
+        if isinstance(pretrained, str):
+            logger = get_root_logger()
+            load_checkpoint(self, pretrained, strict=strict, logger=logger)
+        elif pretrained is None:
+            nn.init.orthogonal_(self.to_rgb.conv.weight)
+            nn.init.orthogonal_(self.noise2feat.weight)
+            # TODO: support user to define init method
+            # xavier_init(self.noise2feat, gain=1, distribution='uniform')
+            # xavier_init(self.to_rgb.conv, gain=1, distribution='uniform')
+        else:
+            raise TypeError("'pretrined' must be a str or None. "
+                            f'But receive {type(pretrained)}.')
+
+
+@MODULES.register_module()
+class ProjDiscriminator(nn.Module):
+    r"""Discriminator for SNGAN / Proj-GAN. The inplementation is refer to
+    https://github.com/pfnet-research/sngan_projection/tree/master/dis_models
+
+    The overall structure of the projection discriminator can be splited
+    to a `from_rgb` layer, a group of ResBlocks, a linear decision layer and
+    a projection layer. To support users to define what kind of layers they
+    To make the discriminator more flexible, we introduce ``from_rgb_cfg``
+    and ``blocks_cfg``, and support users to use their own layers to build the
+    discriminator.
```

Comment:
**Unclear sentences:** (I remove some redundant states.)

The overall structure of the projection discriminator can be split into a `from_rgb` layer, a group of ResBlocks, a linear decision layer, and a projection layer. To support defining custom layers, we introduce ``from_rgb_cfg`` and ``blocks_cfg``.